src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +190,344 @@ public void emit(Event event)
+            break;
+          default:
+            throw new ISE("Unsupported content encoding [%s]", contentEncoding.name());
         }
-        os.write(message);
+      } else {
+        payload = buffer;
+        payloadLength = length;
       }
-      os.write(config.getBatchingStrategy().batchEnd());
-    }
 
-    /**
-     * Splits up messages into batches based on the configured maxBatchSize.
-     *
-     * @param messages list of JSON objects, one per message
-     *
-     * @return sub-lists of "messages"
-     */
-    private List<List<byte[]>> splitIntoBatches(List<byte[]> messages)
-    {
-      final List<List<byte[]>> batches = Lists.newLinkedList();
-      List<byte[]> currentBatch = Lists.newArrayList();
-      int currentBatchBytes = 0;
-
-      for (final byte[] message : messages) {
-        final int batchSizeAfterAddingMessage = config.getBatchingStrategy().batchStart().length
-                                                + currentBatchBytes
-                                                + config.getBatchingStrategy().messageSeparator().length
-                                                + message.length
-                                                + config.getBatchingStrategy().batchEnd().length;
-
-        if (!currentBatch.isEmpty() && batchSizeAfterAddingMessage > config.getMaxBatchSize()) {
-          // Existing batch is full; close it and start a new one.
-          batches.add(currentBatch);
-          currentBatch = Lists.newArrayList();
-          currentBatchBytes = 0;
-        }
 
-        currentBatch.add(message);
-        currentBatchBytes += message.length;
+      request.setContent("application/json", payload, 0, payloadLength);
+
+      if (config.getBasicAuthentication() != null) {
+        final String[] parts = config.getBasicAuthentication().split(":", 2);
+        final String user = parts[0];
+        final String password = parts.length > 0 ? parts[1] : "";
+        request.setBasicAuthentication(user, password);
+      }
+
+      final StatusResponseHolder response = client.go(request, new StatusResponseHandler(Charsets.UTF_8)).get();
+
+      if (response.getStatus().getCode() == 413) {
+        throw new ISE(
+            "Received HTTP status 413 from [%s]. Batch size of [%d] may be too large, "
+            + "try adjusting com.metamx.emitter.http.maxBatchSizeBatch",
+            config.getRecipientBaseUrl(),
+            config.getMaxBatchSize()
+        );
+      }
+
+      if (response.getStatus().getCode() / 100 != 2) {
+        throw new ISE(
+            "Emissions of events not successful[%s], with message[%s].",
+            response.getStatus(),
+            response.getContent().trim()
+        );
       }
+    }
 
-      if (!currentBatch.isEmpty()) {
-        batches.add(currentBatch);
+    GZIPOutputStream acquireGzipOutputStream(int length) throws IOException
+    {
+      if (gzipBaos == null) {
+        gzipBaos = new ZeroCopyByteArrayOutputStream(length);
```

Comment:
`ZeroCopyByteArrayOutputStream` resizes itself if needed (as it extends `ByteArrayOutputStream`)